File: archive/pulumi-ts/Pr5401/lib/network-stack.ts
/**
 * network-stack.ts
 *
 * Creates VPC infrastructure with public and private subnets across multiple AZs.
 */
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface NetworkStackArgs {
  environmentSuffix: string;
  tags?: { [key: string]: string };
}

export class NetworkStack extends pulumi.ComponentResource {
  public readonly vpcId: pulumi.Output<string>;
  public readonly publicSubnetIds: pulumi.Output<string>[];
  public readonly privateSubnetIds: pulumi.Output<string>[];
  public readonly vpcCidr: string;

  constructor(
    name: string,
    args: NetworkStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:network:NetworkStack', name, args, opts);

    const { environmentSuffix, tags } = args;
    this.vpcCidr = '10.0.0.0/16';

    // Create VPC
    const vpc = new aws.ec2.Vpc(
      `vpc-${environmentSuffix}`,
      {
        cidrBlock: this.vpcCidr,
        enableDnsHostnames: true,
        enableDnsSupport: true,
        tags: {
          Name: `vpc-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    this.vpcId = vpc.id;

    // Get availability zones
    const availabilityZones = aws.getAvailabilityZones({
      state: 'available',
    });

    // Create Internet Gateway
    const igw = new aws.ec2.InternetGateway(
      `igw-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        tags: {
          Name: `igw-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create public subnets (2 AZs)
    const publicSubnets: aws.ec2.Subnet[] = [];
    for (let i = 0; i < 2; i++) {
      const subnet = new aws.ec2.Subnet(
        `public-subnet-${i}-${environmentSuffix}`,
        {
          vpcId: vpc.id,
          cidrBlock: `10.0.${i}.0/24`,
          availabilityZone: pulumi
            .output(availabilityZones)
            .apply(azs => azs.names[i]),
          mapPublicIpOnLaunch: true,
          tags: {
            Name: `public-subnet-${i}-${environmentSuffix}`,
            Type: 'public',
            ...tags,
          },
        },
        { parent: this }
      );
      publicSubnets.push(subnet);
    }

    this.publicSubnetIds = publicSubnets.map(s => s.id);

    // Create public route table
    const publicRouteTable = new aws.ec2.RouteTable(
      `public-rt-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        tags: {
          Name: `public-rt-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Route to Internet Gateway
    new aws.ec2.Route(
      `public-route-${environmentSuffix}`,
      {
        routeTableId: publicRouteTable.id,
        destinationCidrBlock: '0.0.0.0/0',
        gatewayId: igw.id,
      },
      { parent: this }
    );

    // Associate public subnets with route table
    publicSubnets.forEach((subnet, i) => {
      new aws.ec2.RouteTableAssociation(
        `public-rta-${i}-${environmentSuffix}`,
        {
          subnetId: subnet.id,
          routeTableId: publicRouteTable.id,
        },
        { parent: this }
      );
    });

    // Create Elastic IP for NAT Gateway
    const natEip = new aws.ec2.Eip(
      `nat-eip-${environmentSuffix}`,
      {
        domain: 'vpc',
        tags: {
          Name: `nat-eip-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create NAT Gateway in first public subnet
    const natGateway = new aws.ec2.NatGateway(
      `nat-gateway-${environmentSuffix}`,
      {
        allocationId: natEip.id,
        subnetId: publicSubnets[0].id,
        tags: {
          Name: `nat-gateway-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create private subnets (2 AZs)
    const privateSubnets: aws.ec2.Subnet[] = [];
    for (let i = 0; i < 2; i++) {
      const subnet = new aws.ec2.Subnet(
        `private-subnet-${i}-${environmentSuffix}`,
        {
          vpcId: vpc.id,
          cidrBlock: `10.0.${10 + i}.0/24`,
          availabilityZone: pulumi
            .output(availabilityZones)
            .apply(azs => azs.names[i]),
          tags: {
            Name: `private-subnet-${i}-${environmentSuffix}`,
            Type: 'private',
            ...tags,
          },
        },
        { parent: this }
      );
      privateSubnets.push(subnet);
    }

    this.privateSubnetIds = privateSubnets.map(s => s.id);

    // Create private route table
    const privateRouteTable = new aws.ec2.RouteTable(
      `private-rt-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        tags: {
          Name: `private-rt-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Route to NAT Gateway
    new aws.ec2.Route(
      `private-route-${environmentSuffix}`,
      {
        routeTableId: privateRouteTable.id,
        destinationCidrBlock: '0.0.0.0/0',
        natGatewayId: natGateway.id,
      },
      { parent: this }
    );

    // Associate private subnets with route table
    privateSubnets.forEach((subnet, i) => {
      new aws.ec2.RouteTableAssociation(
        `private-rta-${i}-${environmentSuffix}`,
        {
          subnetId: subnet.id,
          routeTableId: privateRouteTable.id,
        },
        { parent: this }
      );
    });

    this.registerOutputs({
      vpcId: this.vpcId,
      publicSubnetIds: this.publicSubnetIds,
      privateSubnetIds: this.privateSubnetIds,
    });
  }
}

File: archive/pulumi-java/Pr2276/lib/src/main/java/app/components/NetworkStack.java
package app.components;

import app.config.AppConfig;
import app.utils.TagUtils;
import com.pulumi.aws.AwsFunctions;
import com.pulumi.aws.ec2.InternetGateway;
import com.pulumi.aws.ec2.InternetGatewayArgs;
import com.pulumi.aws.ec2.RouteTable;
import com.pulumi.aws.ec2.RouteTableArgs;
import com.pulumi.aws.ec2.RouteTableAssociation;
import com.pulumi.aws.ec2.RouteTableAssociationArgs;
import com.pulumi.aws.ec2.Subnet;
import com.pulumi.aws.ec2.SubnetArgs;
import com.pulumi.aws.ec2.Vpc;
import com.pulumi.aws.ec2.VpcArgs;
import com.pulumi.aws.ec2.inputs.RouteTableRouteArgs;
import com.pulumi.aws.inputs.GetAvailabilityZonesArgs;
import com.pulumi.aws.outputs.GetAvailabilityZonesResult;
import com.pulumi.core.Output;
import com.pulumi.deployment.InvokeOptions;
import com.pulumi.resources.ComponentResource;
import com.pulumi.resources.ComponentResourceOptions;
import com.pulumi.resources.CustomResourceOptions;

import java.util.Objects;

public class NetworkStack extends ComponentResource {
    private final Output<String> vpcId;
    private final Output<String> publicSubnetPrimaryId;
    private final Output<String> publicSubnetSecondaryId;
    private final Output<String> privateSubnetPrimaryId;
    private final Output<String> privateSubnetSecondaryId;
    private final Output<String> internetGatewayId;
    private final Output<String> publicRouteTableId;

    public NetworkStack(final String name, final AppConfig config, final ComponentResourceOptions options) {
        super("custom:infrastructure:NetworkStack", name, options);

        // Get availability zones
        Output<GetAvailabilityZonesResult> availabilityZones = AwsFunctions.getAvailabilityZones(
                GetAvailabilityZonesArgs.builder()
                        .state("available")
                        .build(),
                InvokeOptions.builder()
                        .provider(Objects.requireNonNull(options.getProvider().orElse(null)))
                        .build()
        );

        // Create VPC
        var vpc = new Vpc(name + "web-hosting-vpc", VpcArgs.builder()
                .cidrBlock(config.getVpcCidrBlock())
                .enableDnsHostnames(true)
                .enableDnsSupport(true)
                .tags(TagUtils.getTagsWithName("WebHosting-VPC", config))
                .build(), CustomResourceOptions.builder()
                .parent(this)
                .build());

        this.vpcId = vpc.id();

        // Create Internet Gateway
        var internetGateway = new InternetGateway(name + "web-hosting-igw",
                InternetGatewayArgs.builder()
                        .vpcId(vpc.id())
                        .tags(TagUtils.getTagsWithName("WebHosting-IGW", config))
                        .build(), CustomResourceOptions.builder()
                .parent(this)
                .build());

        this.internetGatewayId = internetGateway.id();

        // Create Public Subnets
        var publicSubnetPrimary = new Subnet(name + "public-subnet-primary",
                SubnetArgs.builder()
                        .vpcId(vpc.id())
                        .cidrBlock(config.getPublicSubnetPrimaryCidr())
                        .availabilityZone(availabilityZones.applyValue(azs -> azs.names().get(0)))
                        .mapPublicIpOnLaunch(true)
                        .tags(TagUtils.getTagsWithName("Public-Subnet-Primary", config))
                        .build(), CustomResourceOptions.builder()
                .parent(this)
                .build());

        var publicSubnetSecondary = new Subnet(name + "public-subnet-secondary",
                SubnetArgs.builder()
                        .vpcId(vpc.id())
                        .cidrBlock(config.getPublicSubnetSecondaryCidr())
                        .availabilityZone(availabilityZones.applyValue(azs -> azs.names().get(1)))
                        .mapPublicIpOnLaunch(true)
                        .tags(TagUtils.getTagsWithName("Public-Subnet-Secondary", config))
                        .build(), CustomResourceOptions.builder()
                .parent(this)
                .build());

        this.publicSubnetPrimaryId = publicSubnetPrimary.id();
        this.publicSubnetSecondaryId = publicSubnetSecondary.id();

        // Create Private Subnets
        var privateSubnetPrimary = new Subnet(name + "private-subnet-primary",
                SubnetArgs.builder()
                        .vpcId(vpc.id())
                        .cidrBlock(config.getPrivateSubnetPrimaryCidr())
                        .availabilityZone(availabilityZones.applyValue(azs -> azs.names().get(0)))
                        .tags(TagUtils.getTagsWithName("Private-Subnet-Primary", config))
                        .build(), CustomResourceOptions.builder()
                .parent(this)
                .build());

        var privateSubnetSecondary = new Subnet(name + "private-subnet-secondary",
                SubnetArgs.builder()
                        .vpcId(vpc.id())
                        .cidrBlock(config.getPrivateSubnetSecondaryCidr())
                        .availabilityZone(availabilityZones.applyValue(azs -> azs.names().get(1)))
                        .tags(TagUtils.getTagsWithName("Private-Subnet-Secondary", config))
                        .build(), CustomResourceOptions.builder()
                .parent(this)
                .build());

        this.privateSubnetPrimaryId = privateSubnetPrimary.id();
        this.privateSubnetSecondaryId = privateSubnetSecondary.id();

        // Create Route Table for Public Subnets
        var publicRouteTable = new RouteTable(name + "public-route-table",
                RouteTableArgs.builder()
                        .vpcId(vpc.id())
                        .routes(RouteTableRouteArgs.builder()
                                .cidrBlock("0.0.0.0/0")
                                .gatewayId(internetGateway.id())
                                .build())
                        .tags(TagUtils.getTagsWithName("Public-Route-Table", config))
                        .build(), CustomResourceOptions.builder()
                .parent(this)
                .build());

        this.publicRouteTableId = publicRouteTable.id();

        // Associate Route Table with Public Subnets
        new RouteTableAssociation(name + "public-subnet-primary-association",
                RouteTableAssociationArgs.builder()
                        .subnetId(publicSubnetPrimary.id())
                        .routeTableId(publicRouteTable.id())
                        .build(), CustomResourceOptions.builder()
                .parent(this)
                .build());

        new RouteTableAssociation(name + "public-subnet-secondary-association",
                RouteTableAssociationArgs.builder()
                        .subnetId(publicSubnetSecondary.id())
                        .routeTableId(publicRouteTable.id())
                        .build(), CustomResourceOptions.builder()
                .parent(this)
                .build());
    }

    public Output<String> getVpcId() {
        return vpcId;
    }

    public Output<String> getPublicSubnetPrimaryId() {
        return publicSubnetPrimaryId;
    }

    public Output<String> getPublicSubnetSecondaryId() {
        return publicSubnetSecondaryId;
    }

    public Output<String> getPrivateSubnetPrimaryId() {
        return privateSubnetPrimaryId;
    }

    public Output<String> getPrivateSubnetSecondaryId() {
        return privateSubnetSecondaryId;
    }

    public Output<String> getInternetGatewayId() {
        return internetGatewayId;
    }

    public Output<String> getPublicRouteTableId() {
        return publicRouteTableId;
    }
}
